src/ripple/net/HTTPStream.h

```diff
@@ -0,0 +1,148 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NET_HTTPSTREAM_H_INCLUDED
+#define RIPPLE_NET_HTTPSTREAM_H_INCLUDED
+
+#include <ripple/core/Config.h>
+#include <ripple/net/HTTPClientSSLContext.h>
+
+#include <boost/asio/spawn.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+
+namespace ripple {
+
+class HTTPStream
+{
+public:
+    using request = boost::beast::http::request<boost::beast::http::empty_body>;
+    using parser = boost::beast::http::basic_parser<false>;
+
+    virtual ~HTTPStream() = default;
+
+    virtual boost::asio::ip::tcp::socket&
+    getStream() = 0;
+
+    virtual bool
+    connect(
+        std::string& errorOut,
+        std::string const host,
+        std::string const port,
+        boost::asio::yield_context& yield) = 0;
+
+    virtual void
+    asyncWrite(
+        request& req,
+        boost::asio::yield_context& yield,
+        boost::system::error_code& ec) = 0;
+
+    virtual void
+    asyncRead(
+        boost::beast::flat_buffer& buf,
+        parser& p,
+        bool readSome,
+        boost::asio::yield_context& yield,
+        boost::system::error_code& ec) = 0;
+};
+
+class SSLStream : public HTTPStream
+{
+public:
+    SSLStream(
+        Config const& config,
+        boost::asio::io_service::strand& strand,
+        beast::Journal j);
+
+    virtual ~SSLStream() = default;
+
+    boost::asio::ip::tcp::socket&
```

Comment:
It _seems_ like `SSLStream` and `RawStream` provide implementations for `HTTPStream`.  And you might want for people to never directly call methods on, for example, `SSLStream`.  You might instead want them to only call through the base class.  If that's the case, you can consider making all of these `override` methods in `SSLStream` and `RawStream` `private`.  That will force access to alway be through the base class.

It's just an option for you to consider.  Since the pattern is a little unusual you could legitimately choose not to use it.

Reply:
For the moment these classes are called only via `HTTPStream`, but I don't want to force access to them through that interface.